### PR TITLE
[Java] Better log message when failed to invoke task.

### LINKED ISSUE #17704: [Java] Better log message when failed to invoke tasks.
https://github.com/ray-project/ray/blob/master/java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java#L144

`InvocationTargetException` is not enough to be caught. More exceptions may be raised like `IllegageArgumentException`. better to log message with function descriptions, task id and argument types.

### PR DESCRIPTION
**Pull request by jovany-wang**
Catch more broad exceptions when failed to invoke task since `invoke` might throw other exceptions. Better log message will save us away unknown failure reasons.

Close #17704

image

### CODE REVIEW

File: java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java [hunk 1]
```diff
@@ -141,7 +140,7 @@ private void throwIfDependencyFailed(Object arg) {
         } else {
           result = rayFunction.getConstructor().newInstance(args);
         }
-      } catch (InvocationTargetException e) {
+      } catch (Exception e) {
```
Comment (kfstorm) [inline java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:143]:
```suggestion
      } catch (Throwable e) {
```

Comment (kfstorm) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:143]: And please change
```
if (e.getCause() != null) {
    throw e.getCause();
}
```
to
```
if (e instanceof InvocationTargetException && e.getCause() != null) {
    throw e.getCause();
}
```

**Review by kfstorm — COMMENTED** [review]
What other exception types have you encountered?

**Comment by jovany-wang** [comment]
> What other exception types have you encountered?

`IlleagalArgumentException`

File: java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java (hunk 1 above)
Comment (jovany-wang) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:143]: `Throwable` seems too broad to me. It should be handled at outer block.

Comment (WangTaoTheTonic) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:144]: Normally we should not catch such broad exception such as `IOExcepiton` or `Exception`. Is it necessary? or could we narrow it down?

Comment (jovany-wang) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:144]: @WangTaoTheTonic Yes you're right by principle, but here we just log the function and arguments then re-throw it.
And another reason is if we don't catch `Exception`, we have to catch these 4 specific exceptions: InstantiationException, IllegalAccessException, IllegalArgumentException, reflect.InvocationTargetException.

**Comment by kfstorm** [comment]
@jovany-wang Why don't we merge these two log messages?

https://github.com/alipay/ray/blob/b55bc2cece597a51056ba2d071aa51872d406e0a/java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java#L152

https://github.com/alipay/ray/blob/b55bc2cece597a51056ba2d071aa51872d406e0a/java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java#L182

**Comment by jovany-wang** [comment]
> @jovany-wang Why don't we merge these two log messages?
> 
> https://github.com/alipay/ray/blob/b55bc2cece597a51056ba2d071aa51872d406e0a/java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java#L152
> 
> https://github.com/alipay/ray/blob/b55bc2cece597a51056ba2d071aa51872d406e0a/java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java#L182

Sounds good to merge those. Will do that.

File: java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java [hunk 2]
```diff
@@ -127,41 +127,19 @@ private void throwIfDependencyFailed(Object arg) {
...
-              rayFunction,
-              actor,
-              argTypes);
-        }
-
-        if (e.getCause() != null) {
-          throw e.getCause();
-        } else {
-          throw e;
-        }
```
Comment (kfstorm) [inline java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java:163]: Why delete this?

**Comment by jovany-wang** [comment]
Added a test.

File: java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java (hunk 1 above)
Comment (WangTaoTheTonic) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java]: How about we trying to catching these 4 kinds of exceptions if there's only 4?

Comment (jovany-wang) [reply java/runtime/src/main/java/io/ray/runtime/task/TaskExecutor.java]: I removed this line.

File: java/test/src/main/java/io/ray/test/TaskExceptionTest.java [hunk 3]
```diff
@@ -46,4 +47,22 @@ public void testThrowUnserializableExceptionInActorTask() {
...
+  public void testThrowRootExceptionForChainedTasks() {
+    Assert.assertThrows(
+        RayTaskException.class,
+        () -> {
+          try {
+            Ray.task(TaskExceptionTest::hello).remote().get();
+          } catch (Exception e) {
+            throw e.getCause();
+          }
+        });
```
Comment (kfstorm) [inline java/test/src/main/java/io/ray/test/TaskExceptionTest.java:66]:
```suggestion
    RayTaskException ex = Assert.assertThrows(
        RayTaskException.class,
        () -> Ray.task(TaskExceptionTest::hello).remote().get());
    Assert.assertTrue(ex.getCause() instanceof RayTaskException);
```

Comment (jovany-wang) [reply java/test/src/main/java/io/ray/test/TaskExceptionTest.java]: Good.

**Review by kfstorm — APPROVED** [review]

**Review by WangTaoTheTonic — APPROVED** [review]

**Pull request merged**